code/modules/mapping/space_management/zlevel_manager.dm

```diff
@@ -0,0 +1,174 @@
+	var/datum/space_level/S = new z_type(our_z, name, transition_type = linkage, traits = traits)
+	levels_by_name[name] = S
+	z_list += S
+	return our_z
+
+/datum/controller/subsystem/mapping/proc/cut_levels_downto(new_maxz)
+	while(z_list.len > new_maxz)
+		kill_topmost_zlevel()
+
+// Decrements the max z-level by one
+// not normally used, but hey the swapmap loader wanted it
+// Also probably a worse idea than just emptying out levels when needed
+/datum/controller/subsystem/mapping/proc/kill_topmost_zlevel()
+	var/our_z = z_list.len
+	var/datum/space_level/S = get_zlev(our_z)
+	z_list.Cut(our_z, our_z + 1)
+	levels_by_name -= S.name
+	CleanTurfs(S.return_turfs())
+	qdel(S)
+
+/// all turfs must be on the same zlevel
+/datum/controller/subsystem/mapping/proc/CleanTurfs(list/turfs)
+	set waitfor = FALSE
+	LAZYINITLIST(zs_being_cleared)
+	var/turf/first = turfs[1]
+	var/z_key = "[first.z]"
+	var/list/zbc = zs_being_cleared
+	++zbc[z_key]
+	stoplag(1)	//since we KNOW this will take a while, start sleeping so we don't hold up the caller
+	for(var/I in turfs)
+		qdel(I, TRUE)
+		CHECK_TICK
+	if(!--zbc[z_key])
+		zbc -= z_key
+		if(!zbc.len)
+			zs_being_cleared = null
+
+// An internally-used proc used for heap-zlevel management
+/datum/controller/subsystem/mapping/proc/add_new_heap()
+	return add_new_zlevel("Heap level #[z_list.len + 1]", UNAFFECTED, list(BLOCK_TELEPORT, ADMIN_LEVEL), /datum/space_level/heap)
+
+// This is what you can call to allocate a section of space
+// Later, I'll add an argument to let you define the flags on the region
+/datum/controller/subsystem/mapping/proc/allocate_space(width, height)
+	if(width > world.maxx || height > world.maxy)
+		throw EXCEPTION("Too much space requested! \[[width],[height]\]")
+	if(!LAZYLEN(heaps))
+		LAZYADD(heaps, add_new_heap())
+	var/datum/space_level/heap/our_heap
+	var/weve_got_vacancy = 0
+	for(our_heap in heaps)
+		weve_got_vacancy = our_heap.request(width, height)
+		if(weve_got_vacancy)
+			break // We're sticking with the present value of `our_heap` - it's got room
+		// This loop will also run out if no vacancies are found
+
+	if(!weve_got_vacancy)
+		our_heap = add_new_heap()
+		heaps += our_heap
+	return our_heap.allocate(width, height)
+
+/// TRAIT stuff
+
+/datum/controller/subsystem/mapping/proc/check_level_has_trait(z, list/traits)
+	for(var/I in traits)
+		if(check_level_trait(z, I))
+			return TRUE
+	return FALSE
+
+/datum/controller/subsystem/mapping/proc/check_level_has_all_traits(z, list/traits)
+	for(var/I in traits)
+		if(!check_level_trait(z, I))
+			return FALSE
+	return TRUE
+
+/datum/controller/subsystem/mapping/proc/check_level_trait(z, trait)
+	if(!z)
+		return FALSE
+	var/list/trait_list
+	if(z_list)
+		var/datum/space_level/S = get_zlev(z)
+		trait_list = S.flags
+	else
+		var/list/default_map_traits = DEFAULT_MAP_TRAITS
+		trait_list = default_map_traits[z]
+		trait_list = trait_list[DL_ATTRS]
+	return trait_list[trait] ? TRUE : FALSE

```

Comment:
Ah, is this more efficient than `(trait in trait_list)`?

Reply:
Also, is this ternary operator here to scrub the contained value to `TRUE` or `FALSE`?
Would it not be better to do `!!trait_list[trait]` instead?

Reply:
the ternary operator is more immediately obvious in purpose than double negation is so it's no problem imo

Reply:
fair on that point

Reply:
@Crazylemon64 yes, looking up an associative key (list[key]) is MUCH faster than the ```in``` operator, by a landslide.